demoapp/demo_app_functions.py

```diff
@@ -450,4 +466,479 @@ def update_csv(bill_num: str, title: str, summarized_bill: str, category: str, t
+    num_tokens = count_tokens(bill_title, bill_text, mgl_ref, mgl_names, committee_info)
+    
+    return 'small' if num_tokens < 120000 else 'large'
+
+def get_category_tags(categories: List) -> List:
+
+    """
+    This function takes in list of categories and returns tags pertinant to that specifc categories only. 
+
+    Args: 
+
+        categories (List(str)): List of category strings.
+
+    Returns: 
+        List of all tags specific to those of categories.  
+    """
+
+    tags_tuple = itemgetter(*categories)(new_tags_for_bill_dict)
+    category_tags = []
+    for cts in tags_tuple: category_tags += cts
+    return category_tags
+
+def generate_outputs(bill_details: BillDetails, task: str, return_metrics: bool = False) -> tuple:
+
+    '''
+    This function takes in bill details object (bill title, bill text and reference mgl section text) and completes
+    the task of either summarizing or tagging the bill. 
+
+    Arguments: 
+
+    bill_details (BillDetails): Object containing information about the bill - bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+    task (str): It can take 2 values ('summarize','tag')
+    return_metrics (bool): Boolean values indication weather to calculate and return metrics or not - Defaults to False
+
+    Returns: 
+        A tuple of status_code and an LLMResults object containing query, response from the LLM and metrics - if requested. 
+
+        status_code can take these following values {1: Success, -1: Necessary details not found, -2: Invalid task}
+
+    '''
+
+    if not all(hasattr(bill_details, attr) for attr in ("bill_text", 'bill_title', 'mgl_names', 'committee_info')): 
+        return -1, LLMResults()
+
+    llm_call_type = get_llm_call_type(bill_details)   
+
+    if task == 'summarize': 
+
+        query = get_query_for_summarization(bill_details, llm_call_type)
+        return 1, call_llm(bill_details, query, llm_call_type, return_metrics)
+
+    elif task == 'tag':
+
+        query_1 = get_query_for_categorizing(bill_details, llm_call_type)
+        category_response = call_llm(bill_details, query_1, llm_call_type, return_metrics)
+        categories = extract_categories_tags(category_response.response)
+        category_tags = get_category_tags(categories)
+        query_2 = get_query_for_tagging(bill_details, category_tags, llm_call_type)
+        tag_response = call_llm(bill_details, query_2, llm_call_type, return_metrics)
+        tag_response.response = extract_categories_tags(tag_response.response)
+        return 1, tag_response
+
+    return -2, LLMResults()
+
+def extract_categories_tags(response: str) -> list:     
+
+    '''
+    This function parses the response generated by LLM and extracts either categories or tags.
+
+    Args: 
+        response (str): Response string generated by the LLM. Ex: "Tags: ['tag_1', 'tag_2']"
+
+    Returns: 
+        list of either tags or categories. Ex: [tag_1, tag_2]
+
+    '''
+    
+    cats_str = response.split(':')[1]
```

Comment:
This manual parsing is ok, but I think there are LangChain function calling tasks that can help automatically extract structured data from returned outputs. Should we consider using that?

Reply:
I tried using Langchains parsing functions. Given the complexity of having ',' within tags, the parsing functions producing incorrect tags. So, I have resorted to using a hack - asking the model to respond with '#' separated tags.